### PR TITLE
Refactor the kubeadm join command

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
the PR refactors the kubeadm join command and lays the ground for exposing the kubeadm join workflow as phases 

**Which issue(s) this PR fixes**:
Rif #https://github.com/kubernetes/kubeadm/issues/1204

**Special notes for your reviewer**:
The `kubeadm join` command is refactored using the same approach already used for `kubeadm init` in the last cycle: 
1. introduces the `joinOptions` and `joinData` struct 
2. creates a single `run` method that should be progressively migrated to phases

Unit test are provided as well

/sig cluster-lifecycle
/assign @yagonobre 
@kubernetes/sig-cluster-lifecycle-pr-reviews
@timothysc 
@neolit123 
@rosti

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/join.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: just to mention: we didn't expose initOptions in the case of `kubeadm init`, since we don't have `init_test.go`. maybe now that we have init phases we can have some proper unit tests for init.

File: cmd/kubeadm/app/cmd/join.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: i guess the plan for followup PRs is to create a join phase runner and add the phases to it instead of calling them in this Run() function?

one concern here is that we don't have a clear separation between init and join phases - i.e. the init phases are under this namespace:
https://github.com/kubernetes/kubernetes/tree/8288f9e457561e666ee0ab1dbbd2b14c36fbdaed/cmd/kubeadm/app/cmd/phases

maybe we should put them under `phases/init`, `phases/join`?

File: cmd/kubeadm/app/cmd/options/constant.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/options/constant.go]: i think someone submitted a PR for this earlier and it merged.
possible cause for the merge conflict.

File: cmd/kubeadm/app/cmd/join_test.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join_test.go]: sub-tests :+1:

**Review by neolit123 — APPROVED** [review]
@fabriziopandini happy new year!
this PR looks good - i guess this is the first step towards the join phases.
no blocking comments, once the merge conflict is fixed we can LGTM.

**Review by yagonobre — APPROVED** [review]
Thanks @fabriziopandini 
LGTM

File: cmd/kubeadm/app/cmd/join.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/join.go]: @neolit123 yes next pr will add the join phase runner.
I don't think we should have separation between phases, instead when possible we should reuse existing phases into the join workflow, but let's see this once we agree upon a list of phases for this workflow

File: cmd/kubeadm/app/cmd/options/constant.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/options/constant.go]: +1

**Comment by neolit123** [comment]
golint has one complain:
> comment on exported const TokenStr should be of the form "TokenStr ..."

**Review by neolit123 — APPROVED** [review]
tests passed LGTM.
/lgtm
/hold

@fabriziopandini you can cancel the hold if you think this is ready to go, or leave it to gather more review.

**Comment by fabriziopandini** [comment]
@neolit123 this PR is here since 13 days, I don't think that waiting more makes sense
/hold cancel

**Pull request merged**
